components/derive/src/service_interface.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{FnArg, Ident, ItemTrait, TraitItem, TraitItemMethod, Type};
+
+struct ServiceMethodDescriptor {
+    name: Ident,
+    arg_type: Type,
+    id: u32,
+}
+
+// TODO: currently it works only inside exonum crate, also refactor is needed.
+
+pub fn impl_service_interface(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut trait_item: ItemTrait =
+        syn::parse(item.clone()).expect("service_dispatcher can be used on traits only");
+
+    let vec_items = trait_item
+        .items
+        .iter()
+        .filter(|item| {
+            if let TraitItem::Method(_) = item {
```

Comment:
Maybe use `match`?

Reply:
I prefer match too, but I thought that clippy is against this (its not).